lib/Onyx.js

```diff
@@ -1085,41 +1086,48 @@ function set(key, value) {
         return Promise.resolve();
     }
 
-    const valueWithoutNull = removeNullValues(key, value);
-
-    if (valueWithoutNull === null) {
-        return Promise.resolve();
-    }
+    // If the value is null, we remove the key from storage
+    const {value: valueAfterRemoving, wasRemoved} = removeNullValues(key, value);
 
     if (hasPendingMergeForKey(key)) {
-        Logger.logAlert(`Onyx.set() called after Onyx.merge() for key: ${key}. It is recommended to use set() or merge() not both.`);
+        delete mergeQueue[key]
     }
 
-    const hasChanged = cache.hasValueChanged(key, valueWithoutNull);
+    const hasChanged = cache.hasValueChanged(key, valueAfterRemoving);
 
     // This approach prioritizes fast UI changes without waiting for data to be stored in device storage.
-    const updatePromise = broadcastUpdate(key, valueWithoutNull, hasChanged, 'set');
+    const updatePromise = broadcastUpdate(key, valueAfterRemoving, 'set', hasChanged, wasRemoved);
 
-    // If the value has not changed, calling Storage.setItem() would be redundant and a waste of performance, so return early instead.
-    if (!hasChanged) {
+    // If the value has not changed or the key got removed, calling Storage.setItem() would be redundant and a waste of performance, so return early instead.
+    if (!hasChanged || wasRemoved) {
```

Comment:
If we removed keys by setting their values to null, don't we need to update storage with the new object without those keys?

Reply:
`wasRemoved` here only applies when the whole key was set to `null`, not when nested `null` values got removed.

The storage is already being updated, when the `remove` function is called here: https://github.com/margelo/react-native-onyx/blob/ac8faed9890bca04f889f93e0f249191ee6301b5/lib/Onyx.js#L1061

That's why we have to omit calling `Storage.setItem` here again

Reply:
Ooooooh I see. Thank you for explaining.